test/e2e/util_gatewayapi.go

```diff
@@ -0,0 +1,647 @@
+	accepted := false
+	resolvedRefs := false
+
+	// The http route must have at least one parent for which it is successful.
+	// TODO - If it must be successful for all parents, this will need to change.
+	for _, parent := range httproute.Status.Parents {
+		// For each parent conditions should be true for both Accepted and ResolvedRefs
+		for _, condition := range parent.Conditions {
+			switch condition.Type {
+			case string(gwapi.RouteConditionAccepted):
+				acceptedConditionMsg = condition.Message
+				if condition.Status == metav1.ConditionTrue {
+					accepted = true
+				}
+			case string(gwapi.RouteConditionResolvedRefs):
+				resolvedRefConditionMsg = condition.Message
+				if condition.Status == metav1.ConditionTrue {
+					resolvedRefs = true
+				}
+			}
+		}
+		// Check the results for each parent.
+		switch {
+		case !accepted && !resolvedRefs:
+			return fmt.Errorf("httpRoute %s/%s, parent %v/%v neither %v nor %v, last recorded status messages: %s, %s", namespace, name, parent.ParentRef.Namespace, parent.ParentRef.Name, gwapi.RouteConditionAccepted, gwapi.RouteConditionResolvedRefs, acceptedConditionMsg, resolvedRefConditionMsg), nil
+		case !accepted:
+			return fmt.Errorf("httpRoute %s/%s, parent %v/%v not %v, last recorded status message: %s", namespace, name, parent.ParentRef.Namespace, parent.ParentRef.Name, gwapi.RouteConditionAccepted, acceptedConditionMsg), nil
+		case !resolvedRefs:
+			return fmt.Errorf("httpRoute %s/%s, parent %v/%v not %v, last recorded status message: %s", namespace, name, parent.ParentRef.Namespace, parent.ParentRef.Name, gwapi.RouteConditionResolvedRefs, resolvedRefConditionMsg), nil
+		}
+	}
+	t.Logf("httpRoute %s/%s successful", namespace, name)
+	return nil, httproute
+}
+
+// assertHttpRouteConnection checks if the http route of the given name replies successfully,
+// and returns an error if not
+func assertHttpRouteConnection(t *testing.T, hostname string, gateway *gwapi.Gateway) error {
+	t.Helper()
+	domain := ""
+
+	// Create the http client to check the header.
+	client := &http.Client{
+		Timeout: 10 * time.Second,
+		Transport: &http.Transport{
+			TLSClientConfig: &tls.Config{InsecureSkipVerify: true},
+		},
+	}
+
+	// Get gateway listener hostname to use for dnsRecord.
+	if len(gateway.Spec.Listeners) > 0 {
+		if gateway.Spec.Listeners[0].Hostname != nil && len(string(*gateway.Spec.Listeners[0].Hostname)) > 0 {
+			domain = string(*gateway.Spec.Listeners[0].Hostname)
+			if !strings.HasSuffix(domain, ".") {
+				domain = domain + "."
+			}
+		}
+	}
+	// Obtain the standard formatting of the dnsRecord.
+	dnsRecordName := operatorcontroller.GatewayDNSRecordName(gateway, domain)
+
+	// Make sure the DNSRecord is ready to use.
+	if err := assertDNSRecord(t, dnsRecordName); err != nil {
+		return err
+	}
+
+	// Wait and check that the dns name resolves first. Takes a long time.
+	if err := wait.PollImmediate(20*time.Second, 5*time.Minute, func() (bool, error) {
```

Comment:
Might want to consider wrapping this in:
`if net.ParseIP(address) == nil {`
to skip this logic if it's already an IP address.

Only AWS uses DNS names. All other platforms use IP addresses.

Reply:
This is currently only called on AWS, but I will change it for future purposes.